python/activity_stream/data_manager.py

```diff
@@ -0,0 +1,838 @@
+                                                            entity["id"], 
+                                                            "image",
+                                                            load_image=True)
+            self._thumb_map[uid] = {"activity_id": activity_id, 
+                                    "thumbnail_type": self.THUMBNAIL_ENTITY}            
+
+
+    ###########################################################################
+    # sqlite database access methods
+
+    def __init_db(self):
+        """
+        Sets up the database if it doesn't exist.
+        Returns a handle that must be closed.
+        """
+        connection = sqlite3.connect(self._cache_path)
+        
+        # this is to handle unicode properly - make sure that sqlite returns 
+        # str objects for TEXT fields rather than unicode. Note that any unicode
+        # objects that are passed into the database will be automatically
+        # converted to UTF-8 strs, so this text_factory guarantees that any character
+        # representation will work for any language, as long as data is either input
+        # as UTF-8 (byte string) or unicode. And in the latter case, the returned data
+        # will always be unicode.
+        connection.text_factory = str
+                
+        c = connection.cursor()
+        try:
+        
+            # get a list of tables in the current database
+            ret = c.execute("SELECT name FROM main.sqlite_master WHERE type='table';")
+            table_names = [x[0] for x in ret.fetchall()]
+            
+            if len(table_names) == 0:
+                self._bundle.log_debug("Creating schema in sqlite db.")
+                
+                # we have a brand new database. Create all tables and indices
+                c.executescript("""
+                    CREATE TABLE entity (entity_type text, entity_id integer, activity_id integer, created_at datetime);
+                
+                    CREATE TABLE activity (activity_id integer, note_id integer default null, payload blob, created_at datetime);
+                    
+                    CREATE TABLE note (note_id integer, payload blob, created_at datetime);
+                
+                    CREATE INDEX entity_1 ON entity(entity_type, entity_id, created_at);
+                    CREATE INDEX entity_2 ON entity(entity_type, entity_id, activity_id, created_at);
+
+                    CREATE INDEX activity_1 ON activity(activity_id);
+                    CREATE INDEX activity_2 ON activity(activity_id, note_id);
+
+                    CREATE INDEX note_1 ON activity(note_id);
+                    """)
+                connection.commit()
+        except:
+            connection.close()
+            c = None
+            raise
+
+        finally:
+            if c:
+                c.close()
+
+        return connection
+ 
+    def __get_note_thread_data(self, note_id):
```

Comment:
These db methods all have the same try/except/finally structure. Perhaps a context manager to get the cursor and do the cleanup for you?